### PR TITLE
[WIP] a hook sidecar to manipulate disk driver cache modes

### PR DESCRIPTION
**Pull request by enp0s3**
Signed-off-by: enp0s3 <ibezukh@redhat.com>

**What this PR does / why we need it**:
In order to test some IO performance regarding block devices and page cache usage
we would like to use various cache modes for the disk driver, w/o making changes to the product
code. Thus the following hook sidecar is suggested

**Which issue(s) this PR fixes** #
https://bugzilla.redhat.com/show_bug.cgi?id=2162140

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by alicefr** [comment]
@enp0s3 would be possible to keep this sidecar generic for disk parameters? I think it could be used also for setting iotune parameters. This is a feature that comes up regularly (https://github.com/kubevirt/kubevirt/issues/5444 or https://github.com/kubevirt/kubevirt/pull/7605). The iotune parameters can be done in a following PR, but we could at least keep the naming generic. WDYT?

File: cmd/disk-cache-hook/disk-cache-hook.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+const (
+	onDefineDomainLoggingMessage = "Hook's OnDefineDomain callback method has been called"
+	hookName                     = "disk-driver-cache-hook"
+	version                      = "v1alpha1"
+)
+
+type infoServer struct{}
+
+func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*hooksInfo.InfoResult, error) {
+	log.Log.Info("Hook's Info method has been called")
```
Comment (0xFelix) [inline cmd/disk-cache-hook/disk-cache-hook.go:49]: Make this a const too?

File: cmd/disk-cache-hook/disk-cache-hook.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		},
+	}, nil
+}
+
+type hookServer struct {
+	cacheType string
+}
+
+func (u hookServer) OnDefineDomain(ctx context.Context, params *hooksV1alpha1.OnDefineDomainParams) (*hooksV1alpha1.OnDefineDomainResult, error) {
+	log.Log.Info(onDefineDomainLoggingMessage)
```
Comment (0xFelix) [inline cmd/disk-cache-hook/disk-cache-hook.go:70]: Or inline the message here to be consistent?

File: cmd/disk-cache-hook/disk-cache-hook.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &hooksV1alpha1.OnDefineDomainResult{
+		DomainXML: newDomainXML,
+	}, nil
+}
+
+func onDefineDomain(domainSpec domainSchema.DomainSpec, cacheType string) ([]byte, error) {
```
Comment (0xFelix) [inline cmd/disk-cache-hook/disk-cache-hook.go:89]: Might rename this to something more descriptive like `setDomainDisksCacheType`?

File: cmd/disk-cache-hook/disk-cache-hook.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		panic(err)
+	}
+
+	return newDomainXML, nil
+}
+
+func main() {
+	log.InitializeLogging(hookName)
+
+	var cacheType string
```
Comment (0xFelix) [inline cmd/disk-cache-hook/disk-cache-hook.go:118]: Might make sense to verify the value of `cacheType`?

**Review by 0xFelix — COMMENTED** [review]
Added some comments. Can you add the binary to `.gitignore`?

**Comment by enp0s3** [comment]
@rmohr Hi. Yes exactly, was also thinking to take it out.

**Pull request closed without merging**